[CORE-290] Use/Store consolidated spend report query results in workspace spend report table

samanehsan opened the pull request:
Ticket: https://broadworkbench.atlassian.net/browse/CORE-290

Changes:
- Query the workspace spend report table in Rawls for rows that match the given google project ids, start date and end date
- If there is a result for all project ids (a full cache hit), use those results and filter out any records where `isDataAvailable = false`
- Otherwise:
  - Run the bigQuery job and store those results in the workspace spend report table
  - For every google project id that does not have any query results, store a "dummy record" for that project id, start and end date in the database with `isDataAvailable = false` and no cost data. Otherwise subsequent calls will get a cache miss.

To-Do:
- [x] Handle the case where workspace spend is unavailable
- [x] Handle credits (need to add credits per category to the db table first) 
- [ ] Make matching by start/end more flexible (just check M/D/Y and not the exact timestamp)?
- [ ] Delete records older than X days (could make this a follow-on PR)
- [x] Tests!

---

**PR checklist**

- [x] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should publish a new official `rawls-model` and perform the corresponding dependency updates as specified in the [README](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model):
  - [ ] in the automation subdirectory
  - [ ] in workbench-libs
  - [ ] in firecloud-orchestration
- [ ] Get two thumbsworth of PR review
- [x] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

samanehsan committed:
Add WorkspaceSpendReport model and DB access

samanehsan committed:
Store consolidated spend query results in db

samanehsan committed:
Use cached results or re-query and store

samanehsan committed:
Include workspace name when returning cached spend info

samanehsan committed:
Handle workspaces with no billing info

The results should be stored in the Rawls DB
in order to get a full cache hit, but should
not be returned in the API. The UI will display
"N/A" for the workspaces that are not returned.

samanehsan committed:
Scala format

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-290-use-spend-report-table

samanehsan committed:
Add credit and currency fields

samanehsan committed:
Scalafmt

samanehsan committed:
Fix SpendReportingServiceSpec constructors

samanehsan committed:
Fix inserting N/A workspace spend records

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-290-use-spend-report-table

samanehsan committed:
Fix tests

samanehsan committed:
Fix more test errors

samanehsan committed:
Handle infrastructure cost case

samanehsan committed:
Fix contract tests

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-290-use-spend-report-table

samanehsan committed:
Add tests for new files

samanehsan committed:
Fix filtering spend reports by timestamps

samanehsan committed:
Fix SpendReportingServiceSpec mocking

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,304 @@
+      workspaceSpendReport.id,
+      workspaceSpendReport.googleProjectId,
+      new Timestamp(workspaceSpendReport.reportStartDate.toInstant(ZoneOffset.UTC).toEpochMilli),
+      new Timestamp(workspaceSpendReport.reportEndDate.toInstant(ZoneOffset.UTC).toEpochMilli),
```
can you simplify this to `Timestamp.valueOf(workspaceSpendReport. reportEndDate)`? Is there a complication with the timezone?

samanehsan replied:
Yeah I noticed a timezone difference, but let me try this again...maybe the issue was with my test inputs?

samanehsan replied:
I was doing some manual testing and noticed an issue in how I'm handling these timestamps. For instance if the end datetime is `2025-02-10T23:59:59.999Z`, it gets converted to the timestamp before inserting it into the db `2025-02-10 23:59:59.999` but it gets stored as `2025-02-11 00:00:00` in mysql 🤔

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,304 @@
+      record.id,
+      record.googleProjectId,
+      LocalDateTime.ofInstant(record.reportStartDate.toInstant, ZoneOffset.UTC),
+      LocalDateTime.ofInstant(record.reportEndDate.toInstant, ZoneOffset.UTC),
```
Similar comment; can you simplify this to `record.reportEndDate.toLocalDateTime`?

Edit: there's a bunch of other conversions elsewhere - this comment would apply generally to all of them

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,304 @@
+    dateTimeOpt match {
+      case Some(dateTime) =>
+        val instant = dateTime.toInstant
+        LocalDateTime.ofInstant(java.time.Instant.ofEpochMilli(dateTime.getMillis), ZoneId.systemDefault())
```
can this be just `dateTime.toLocalDateTime`?

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,304 @@
+        cost1 <- totalCompute
+        cost2 <- totalStorage
+        cost3 <- otherSpend
+      } yield cost1 + cost2 + cost3
```
You could use a one-liner here:
```scala
val cost: Option[Float] = Option(Seq(totalCompute, totalStorage, otherSpend).flatten.sum)
```
your call on whichever is more readable

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -571,28 +574,87 @@ class SpendReportingService(
+      reports.map(r => r.spendSummary.googleProjectId).collect { case Some(project) => project.toString() }.toSeq
+    val filteredProjectNames: Map[GoogleProjectId, WorkspaceName] = projectNames.filter { case (id, _) =>
+      !googleProjectsWithReports.contains(id.toString())
     }
```
This code creates `googleProjectsWithReports ` as a `Seq`, then checks a bunch of other values to see if they exist in that `Seq`. When doing lookups like that, a `Set` is much more performant. At this scale it might not matter, but it's a good practice. Additionally, once you have a `Set` you can use its utility methods to help with diff/intersection/union/etc.

I might rewrite this code as:
```scala
    // extract all google project ids from the supplied SpendReportingResults
    val googleProjectsWithReports: Set[GoogleProjectId] =
      results                                     // Iterable[Option[SpendReportingResults]]
        .flatten                                  // Iterable[SpendReportingResults]
        .flatMap(_.spendSummary.googleProjectId)  // Iterable[GoogleProject]
        .map(p => GoogleProjectId(p.value))       // Iterable[GoogleProjectId]
        .toSet                                    // Set[GoogleProjectId]
    // find all supplied project names which are not in the SpendReportingResults projects
    val missingProjectIds: Set[GoogleProjectId] = projectNames.keySet diff googleProjectsWithReports
```

calypsomatic commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,304 @@
+      case List() => throw RawlsExceptionWithErrorReport(StatusCodes.NotFound, "No currencies found for spend data")
+    }
+
+    val all = records.map { record =>
```
This chunk of code seems pretty similar to what's in `SpendReportingService`.  Is there a way it can be reused?

davidangb replied:
+1, would love to see alignment/reuse here

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,304 @@
+      spendDetail.spendData.map {
+        var totalStorage: Option[Float] = None
+        var totalCompute: Option[Float] = None
+        var otherSpend: Option[Float] = None
+        var storageCredits: Option[Float] = None
+        var computeCredits: Option[Float] = None
+        var otherCredits: Option[Float] = None
+        spendData =>
+          spendData.subAggregation.get.spendData.foreach { categorySpend =>
+            val categoryCost = Some(categorySpend.cost.toFloat)
+            val categoryCredits = Some(categorySpend.credits.toFloat)
+            categorySpend.category match {
+              case Some(TerraSpendCategories.Storage) =>
+                totalStorage = categoryCost
+                storageCredits = categoryCredits
+              case Some(TerraSpendCategories.Compute) =>
+                totalCompute = categoryCost
+                computeCredits = categoryCredits
+              case Some(TerraSpendCategories.Other) =>
+                otherSpend = categoryCost
+                otherCredits = categoryCredits
+              // This is not included in the consolidated spend report
+              case Some(TerraSpendCategories.WorkspaceInfrastructure) =>
+              case None                                               =>
+            }
+          }
+          WorkspaceSpendReport.newWorkspaceSpendReport(
+            spendData.googleProjectId.get.value,
+            convertJodaToJava(summary.startTime),
+            convertJodaToJava(summary.endTime),
+            summary.currency,
+            isDataAvailable = true,
+            totalCompute,
+            totalStorage,
+            otherSpend,
+            computeCredits,
+            storageCredits,
+            otherCredits
+          )
+      }
```
Use of vars in Scala is generally a code smell. You can avoid that by arranging the incoming data into a more friendly format, then extracting from that data (think about the ARRANGE/ACT/ASSERT pattern for tests, without ASSERT). I would do it something like this:
```scala
      spendDetail.spendData.map { spendData =>
        // create a map of category->spend
        val spendCategoryMap: Map[Option[TerraSpendCategories.TerraSpendCategory], SpendReportingForDateRange] =
          spendData.subAggregation
            .flatMap(_.spendData)
            .map(datedReport => datedReport.category -> datedReport)
            .toMap
        // helper to look up the cost and credits for a given category
        def getCategoryCost(category: TerraSpendCategories.TerraSpendCategory): (Option[Float], Option[Float]) =
          spendCategoryMap
            .get(Some(category))
            .map { categorySpend =>
              (Option(categorySpend.cost.toFloat), Option(categorySpend.credits.toFloat))
            }
            .getOrElse((None, None))

        // get cost and credits for the categories we care about
        val (totalStorage, storageCredits) = getCategoryCost(TerraSpendCategories.Storage)
        val (totalCompute, computeCredits) = getCategoryCost(TerraSpendCategories.Compute)
        val (otherSpend, otherCredits) = getCategoryCost(TerraSpendCategories.Other)

        WorkspaceSpendReport.newWorkspaceSpendReport(
          spendData.googleProjectId.get.value,
          convertJodaToJava(summary.startTime),
          convertJodaToJava(summary.endTime),
          summary.currency,
          isDataAvailable = true,
          totalCompute,
          totalStorage,
          otherSpend,
          computeCredits,
          storageCredits,
          otherCredits
        )
      }
```

samanehsan replied:
I'm getting an error on this line `.flatMap(_.spendData)` that I can't quite figure out how to fix 🤔 
```
type mismatch;
 found   : Seq[org.broadinstitute.dsde.rawls.model.SpendReportingForDateRange]
 required: Option[?]

davidangb replied:
Yup, there was definitely an error in there … does this work better?
```scala
        val spendCategoryMap: Map[Option[TerraSpendCategories.TerraSpendCategory], SpendReportingForDateRange] =
          spendData.subAggregation
            .map(_.spendData)
            .getOrElse(Seq())
            .map(datedReport => datedReport.category -> datedReport)
            .toMap
```

samanehsan replied:
That fixed it, thanks!!

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -571,28 +574,87 @@ class SpendReportingService(
           return Future.successful(None)
         }
         projectNames: Map[GoogleProjectId, WorkspaceName] = billingMap.values.flatten.toMap
+        projectIds = projectNames.keySet.map(x => x.value)
```
nit: you don't need an `x` variable here:
```suggestion
        projectIds = projectNames.keySet.map(_.value)
```

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -571,28 +574,87 @@ class SpendReportingService(
+          val hasDataAvailable = cacheResults.filter(cached => cached.isDataAvailable)
+          return Future.successful(
+            Some(workspaceSpendReportRepository.toSpendReportResults(hasDataAvailable, projectNames))
+          )
```
early `return`s are non-idiomatic Scala … fun with Scala! Because this for-comprehension is already pretty big, I'd look to solve this by breaking out everything after the `return` into a separate method. Then, you could do something like:
```scala
        combinedResults <-
          if (cacheResults.size == projectIds.size) {
            val hasDataAvailable = cacheResults.filter(cached => cached.isDataAvailable)
            Future.successful(
              Some(workspaceSpendReportRepository.toSpendReportResults(hasDataAvailable, projectNames))
            )
          } else {
            getUncachedSpendForAllWorkspaces(tableToProjectIdsMap,
                                             projectNames,
                                             start,
                                             end,
                                             pageSize,
                                             offset,
                                             childContext
            )
          }
      } yield combinedResults

      private def getUncachedSpendForAllWorkspaces( …
```

samanehsan replied:
That looks much better!!

davidangb reviewed (commented):
Some early feedback for you to think about, mostly about Scala idioms

samanehsan committed:
Code review comments - spend reporting service

samanehsan committed:
Updates to WorkspaceSpendReportComponent

samanehsan committed:
Fix tests and scalafmt

samanehsan committed:
Add SpendReportingServiceSpec test

samanehsan committed:
Merge branch 'develop' of github.com:broadinstitute/rawls into se/CORE-290-use-spend-report-table

samanehsan committed:
Fix merge conflicts

samanehsan committed:
Fix Rawls billing project model

samanehsan committed:
Fix mocks and getting currency

samanehsan committed:
Fix service tests

samanehsan committed:
Simplify timestamp conversion

samanehsan committed:
Bug fixes

- Fix datetime precision
- Fix inserting records with no spend report data

samanehsan commented:
The test that's failing is flaky `SpendReportingServiceSpec.getSpendForAllWorkspaces should get the spend report from multiple billing projects` and I'm trying to figure out why the results are sometimes none 🤔

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,255 @@
+          SpendReportUtils.toBigDecimal(record.otherCredits, currencyCode).toString(),
+          currencyCode.toString,
+          category = Option(TerraSpendCategories.Other)
+        ),
```
the `toBigDecimal` will turn `None` into `0.00` … is that correct here? That means we will never report `N/A`, we'll report `0`. This may be what we want; I am not sure of the requirements

samanehsan replied:
I think that's okay (assuming I'm understanding the behavior correctly)? It looks like the UI displays `N/A` on the consolidated spend report if a workspace is not included in the API response from Rawls. When we initially query the data, this should insert an empty record for the workspace with isDataAvailable=False. When the cached results are returned, only records with isDataAvailable=True are included in the API response.

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportUtils.scala:
```diff
@@ -0,0 +1,46 @@
+          s"Inconsistent currencies found while aggregating spend data: $head and ${tail.head} cannot be combined"
+        )
+      case List() => throw RawlsExceptionWithErrorReport(StatusCodes.NotFound, "No currencies found for spend data")
+    }
```
this also duplicates code in `SpendReportingService.extractSpendReportingResults`, would be good to consolidate

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -617,26 +615,88 @@ class SpendReportingService(
+        Option.empty,
+        Option.empty,
+        Option.empty
+      )
```
You could make a new method `WorkspaceSpendReport.newEmptyReport` or whatever you want to name it, and include all the defaults so you don't have to write them all out here. Optional; your call

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingService.scala:
```diff
@@ -617,26 +615,88 @@ class SpendReportingService(
+                }
+              queryResults.flatMap { res =>
+                Future.successful(insertRecordsWithMissingSpendData(res, workspaceNamesByProjectId, start, end))
+              }
```
Two suggestions here:

1. you can simplify this with `.map` instead of `.flatMap`:
```suggestion
              queryResults.map { res =>
                insertRecordsWithMissingSpendData(res, workspaceNamesByProjectId, start, end)
              }

```

2. I think this should catch and ignore any errors that happen when saving records to the cache. If we hit an error here, we can still deliver results to the end user; we'll just fail to cache those results. There is a race condition which makes errors here likely: since the calls to `workspaceSpendReportRepository.getWorkspaceSpendReports` and `workspaceSpendReportRepository.insertWorkspaceSpendReport` are each in their own separate transaction, if multiple users hit an uncached consolidated spend report at the same time, both could see no records in cache, and then both would try to write those records to cache. One of those writes will fail due to the unique constraints on the cache table.

davidangb commented on core/src/test/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingServiceSpec.scala:
```diff
@@ -1761,6 +1793,186 @@ class SpendReportingServiceSpec extends AnyFlatSpecLike with Matchers with Mocki
+    )
+    spendSummary.endTime.get.toString(ISODateTimeFormat.date()) shouldBe to.toString(ISODateTimeFormat.date())
+    verifyNoInteractions(bigQueryService)
+  }
```
Suggestions for additional unit tests: 
1. What happens when cached results exist for the requested dates, but only cover a subset of the requested project?
2. What happens when cached results exist for all the requested projects, but for the wrong dates?
3. A test that verifies that after we get results from BigQuery, we persist those results to the cache. This may need to use an `eventually` since the cache-writes are fire-and-forget.

davidangb reviewed (commented):
more comments inline, this is getting close!

samanehsan committed:
Additional code review changes

samanehsan committed:
Add more unit tests

samanehsan committed:
Add unit tests for insertRecordsWithMissingSpendData

samanehsan committed:
Consolidate creating spend reports from BQ rows/DB records

samanehsan committed:
Merge branch 'develop' into se/CORE-290-use-spend-report-table

samanehsan commented:
@davidangb do you have any idea why SpendReportingServiceSpec.getSpendForAllWorkspaces `should get the spend report from multiple billing projects` and `should get the spend report from multiple billing projects when cached reports are from different dates` are flaky? 🤔 Sometimes the result of `getSpendForAllWorkspaces` in those tests is None

davidangb commented on core/src/test/scala/org/broadinstitute/dsde/rawls/spendreporting/SpendReportingServiceSpec.scala:
```diff
@@ -1440,92 +1574,75 @@ class SpendReportingServiceSpec extends AnyFlatSpecLike with Matchers with Mocki
+    when(job2.getQueryResults(any())).thenReturn(createTableResult(table2))
+    when(job2.getStatistics).thenReturn(mock[JobStatistics.QueryStatistics](RETURNS_SMART_NULLS))
+    when(job2.waitFor()).thenReturn(job2)
+    when(bigQueryService.runJob(any(), any())).thenReturn(IO(job1), IO(job2))
```
@samanehsan I'm pretty sure this is the cause of flakiness. this always returns job1 and job2 in that order, but they are not always executed in that order, so sometimes execution fails.

davidangb committed:
fix flaky test? (#3221)

samanehsan committed:
Scala fmt

samanehsan committed:
Fix another flaky test

samanehsan commented:
@calypsomatic @davidangb let me know if there are other changes you'd like to see, or if I can answer any questions!

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala:
```diff
@@ -0,0 +1,291 @@
+                               otherCredits: Float,
+                               projectNames: Map[GoogleProjectId, WorkspaceName]
+  ): SpendReportingForDateRange = {
+    val currencyString = currency
+    val currencyCode = Currency.getInstance(currencyString)
```
nit:
```suggestion
    val currencyCode = Currency.getInstance(currency)
```

davidangb reviewed (approved):
one tiny nitpick inline. Nice job getting this through! I am very curious to see how it performs in practice.

samanehsan committed:
Update core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceSpendReportComponent.scala

Co-authored-by: David An <davidan@broadinstitute.org>

samanehsan commented:
@davidangb I just did a final round of manual testing as a sanity check and found that adding a jobId for testing is causing this error:

```Error fetching results from BigQuery: Invalid job ID "broad-gcp-billing.gcp_billing_export_views.gcp_billing_export_v1_001AC2_2B914D_822931". Job IDs must be alphanumeric (plus underscores and dashes) and must be at most 1024 characters long```

What do you think about using `jobId.hashCode.toString` to create a hash of the job id?

davidangb commented:
> What do you think about using jobId.hashCode.toString to create a hash of the job id?

@samanehsan I worry just a little bit about hash collisions when using .hashCode, even though it's probably just fine. Maybe base64-encode the value via `Base64.getEncoder.encodeToString(jobId.getBytes).take(1023)`? If a hash collision is only a problem for unit tests it's no big deal … I wonder if BigQuery would have any kind of problem if it saw duplicate job ids … probably not?

samanehsan commented:
What about making the jobId param an optional and only setting the job id if it exists? Then we can just change this behavior for the tests.

samanehsan commented:
@davidangb I tested the behavior, and re-using the same job id (which would be the spend export table name) in BigQuery returns the same query results 🤯: That is not what I expected, based on https://cloud.google.com/bigquery/docs/managing-jobs#repeat_jobs

Specifically I ran a query to get workspace spend reports from this year, then when I changed the report dates to last November it re-queried BigQuery but returned the same results (including a workspace that didn't even exist until January)

samanehsan committed:
Use job labels for unit test mocking

Using the same BigQuery job id for different
jobs causes the same results to get returned.

samanehsan committed:
Merge branch 'se/CORE-290-use-spend-report-table' of github.com:broadinstitute/rawls into se/CORE-290-use-spend-report-table

samanehsan committed:
Remove job labels

Table name is already present in the SQL query that
is part of the job query config so the tests can use
that instead of adding a label

samanehsan committed:
Merge branch 'develop' into se/CORE-290-use-spend-report-table
